New FAQ Page

kvnkho opened the pull request:
## Summary

This is a FAQ for the most commonly asked questions in the Prefect Slack channel. I will make some edits to the Core FAQ to move one or two questions over also. 

## Changes
A new FAQ page in the docs

## Importance
It provides one page for users to view the most frequently asked questions.

## Checklist
<!-- PRs will not be reviewed unless these boxes are checked -->

This PR:

- [ ] adds new tests (if appropriate)
- [ ] adds a change file in the `changes/` directory (if appropriate)
- [ ] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

madkinsz commented:
If feasible, can we have these patterns in their relevant documentation sections and just link to them from here? cc @tpdorsey for thoughts.

tpdorsey commented on docs/orchestration/faq/faq.md:
````diff
@@ -0,0 +1,228 @@
+	do_something(start_date)
+```
+
+The default Parameter value is evaluated during registration time, so the start_date in the example above will be fixed to the registration date. In order to add dynamicism, using a task is needed to defer the execution of date.today() to runtime. For example:
````
```suggestion
The default Parameter value is evaluated during registration time, so the `start_date` in the example above will be fixed to the registration date. In order to add dynamicism, using a task is needed to defer the execution of `date.today()` to runtime. For example:
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### How can I set a custom flow run name?
+
+The flow run name can not be set in advance, but it can be changed using the [RenameFlowRun](api/latest/tasks/prefect.html#renameflowrun) task after the flow run has been created. You can use this task inside the Flow block, or through a flow-level state handler. When calling the task from a state handler, make sure to call the task’s  `.run()` method.
```
```suggestion
The flow run name cannot be set in advance, but it can be changed using the [RenameFlowRun](api/latest/tasks/prefect.html#renameflowrun) task after the flow run has been created. You can use this task inside the flow block, or through a flow-level state handler. When calling the task from a state handler, make sure to call the task’s  `.run()` method.
```

tpdorsey replied:
The RenameFlowRun URL may need to be adjusted as well

tpdorsey commented on docs/orchestration/faq/faq.md:
````diff
@@ -0,0 +1,228 @@
+    first_task()
+```
+
+### How do I add things to the prefect context so I can reuse them in other tasks?
````
```suggestion
### How do I add things to the Prefect context so I can reuse them in other tasks?
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+The Prefect context is not meant to be mutable so adding an item to it inside a task will not take effect in downstream tasks. Often, a much better approach is to use a Parameter to contain a value that will be used or manipulated by tasks.
+
+### How do I pull Cloud secrets instead of using Local Secrets?
```
```suggestion
### How do I pull Cloud secrets instead of using local secrets?
```

tpdorsey commented on docs/orchestration/faq/faq.md:
````diff
@@ -0,0 +1,228 @@
+```bash
+export PREFECT__CLOUD__USE_LOCAL_SECRETS=false
+```
+or you can use the `config.toml` setting
````
```suggestion

Or you can use the `config.toml` setting:

```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### [REVIEW] Can I run a Docker agent in a container?
+
+In general, the Docker agent is supposed to run in a local process (rather than in a docker container), and this local process is a layer between Prefect backend and a Docker Daemon.
```
```suggestion
In general, the Docker agent is supposed to run in a local process (rather than in a Docker container), and this local process is a layer between Prefect backend and a Docker daemon.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+This agent polls the API for new flow runs, and if there are new scheduled runs, it then creates new flow runs and deploys those as Docker containers on the same machine as the agent.
+
+When the Docker agent is running within a container itself (rather than a local process), your flow runs end up deployed as containers, but not as individual containers, but rather within the agent container. You effectively have a single agent container spinning up new containers within itself (docker in docker), which may have many unintended consequences such as issues with scale and resource  ...
```
```suggestion
When the Docker agent is running within a container itself (rather than a local process), your flow runs end up deployed as containers &mdash; not as individual containers, but rather within the agent container. You effectively have a single agent container spinning up new containers within itself (Docker in Docker), which may have many unintended consequences such as issues with scale and resource utilization.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+If you want more environment isolation for this agent process, you can run it within a virtual environment.
+
+And if you have a strict requirement that every process must run in a container, consider using the [KubernetesAgent](orchestration/agents/kubernetes.html#requirements) instead.
```
```suggestion
And if you have a strict requirement that every process must run in a container, consider using the [KubernetesAgent](/orchestration/agents/kubernetes.html#requirements) instead.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### How can I change the number of DaskExecutor workers based on a Parameter value?
+
+You can pass a Callable to the DaskExecutor that sizes the DaskExecutor dynamically at runtime.
```
```suggestion
You can pass a callable to the DaskExecutor that sizes the DaskExecutor dynamically at runtime.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+def my_task(condition: bool=True):
+  if condition:
+	  raise FAIL('this task failed')
+	else:
```
```suggestion
  else:
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+  if condition:
+	  raise FAIL('this task failed')
+	else:
+		return "all is good"
```
```suggestion
    return "all is good"
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+If it is working, you might need to start the Server with the expose flag as follows: `prefect server start --expose`. This will allow outside connections. You can find more information in this [Github issue](https://github.com/PrefectHQ/prefect/issues/4963).
+
+### Why does Prefect mark Flow runs with `No heartbeat detected from the remote task; marking the run as failed`?
```
```suggestion
### Why does Prefect mark flow runs with `No heartbeat detected from the remote task; marking the run as failed`?
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### Why does Prefect mark Flow runs with `No heartbeat detected from the remote task; marking the run as failed`?
+
+Prefect Flows have heartbeats that signal to Prefect Cloud that your Flow is alive. If Prefect didn’t have heartbeats, Flows that lose communication and die would permanently be shown with a Running state in the UI. Most of the time, we have seen “no heartbeat detected” to be a sign of memory issues. In version 0.15.4 and above, additional logging has been added that propagates the real error in  ...
```
```suggestion
Prefect flows have heartbeats that signal to Prefect Cloud that your flow is alive. If Prefect didn’t have heartbeats, flows that lose communication and die would be shown permanently with a Running state in the UI. Most of the time, we have seen “no heartbeat detected” to be a sign of memory issues. In version 0.15.4 and above, additional logging has been added that propagates the real error in the event this error happens. 
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+Prefect Flows have heartbeats that signal to Prefect Cloud that your Flow is alive. If Prefect didn’t have heartbeats, Flows that lose communication and die would permanently be shown with a Running state in the UI. Most of the time, we have seen “no heartbeat detected” to be a sign of memory issues. In version 0.15.4 and above, additional logging has been added that propagates the real error in  ...
+
+You can [configure heartbeats to use threads instead of processes.](orchestration/concepts/services.html#heartbeat-configuration) This has proven to be more stable for a lot of users. In the event that the Flow continues to fail after switching to using threads, some users have had success in making the failing task a subflow, and then turning off heartbeats for the subflow.
```
```suggestion
You can [configure heartbeats to use threads instead of processes.](/orchestration/concepts/services.html#heartbeat-configuration) This has proven to be more stable for many users. In the event that the flow continues to fail after switching to using threads, some users have had success in making the failing task a subflow, and then turning off heartbeats for the subflow.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### How do I attach other loggers to the Prefect logger?
+
+The logging docs show how [here](core/concepts/logging.html#extra-loggers)
```
```suggestion
The logging docs show how [here](/core/concepts/logging.html#extra-loggers).
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### Why do I get a `ModuleNotFoundError` with my home directory path?
+
+When a flow is registered, Prefect stores the location of it in [Storage](orchestration/flow_config/storage.html) (`GitHub`, `S3`, `Docker`, etc.). During a flow run execution, Prefect pulls the flow from the storage location and runs it. If users don’t specify any storage, it defaults to a `Local` storage, which is a serialized version of the flow stored in the `~/.prefect/flows` folder. At runt ...
```
```suggestion
When a flow is registered, Prefect stores the location of it in [Storage](/orchestration/flow_config/storage.html) (`GitHub`, `S3`, `Docker`, etc.). During a flow run execution, Prefect pulls the flow from the storage location and runs it. If users don’t specify any storage, it defaults to a `Local` storage, which is a serialized version of the flow stored in the `~/.prefect/flows` folder. At runtime, the flow is retrieved from this file.
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+- Check that there is a healthy agent that can pick up the Flow.
+- Make sure that the RunConfig takes in a `List[str]` as opposed to just a string. Setting `labels="prod"` will be treated as four labels `(["p", "r", "o", "d"])` because strings are iterable in Python.
+- Flows with no labels are picked up by Agents with no labels only.
+- There is a [default label on the LocalAgent that contains the hostname](orchestration/agents/local.html#labels).
```
```suggestion
- There is a [default label on the LocalAgent that contains the hostname](/orchestration/agents/local.html#labels).
```

tpdorsey commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+- Make sure that the RunConfig takes in a `List[str]` as opposed to just a string. Setting `labels="prod"` will be treated as four labels `(["p", "r", "o", "d"])` because strings are iterable in Python.
+- Flows with no labels are picked up by Agents with no labels only.
+- There is a [default label on the LocalAgent that contains the hostname](orchestration/agents/local.html#labels).
+- There is a [default label on the default Local Storage](orchestration/flow_config/storage.html#local).
```
```suggestion
- There is a [default label on the default Local Storage](/orchestration/flow_config/storage.html#local).
```

tpdorsey commented:
> > If feasible, can we have these patterns in their relevant documentation sections and just link to them from here? cc @tpdorsey for thoughts.
> 
> I don't mind doing that. The intention was someone asks a question in support and we link them to this one page. I understand the sentiment though that the question might not even be asked in the first place if the info were in the docs. I defer to @tpdorsey to decide. I suspect there will be some that don't fit anywhere. For those, we can leave the answer typed out for now. For any that makes sense somewhere else, we can certainly do that.

Suggest copying relevant pattern examples to relevant docs as appropriate _and_ link to the relevant docs as well. Makes the correct practice easier to find depending on how the user is accessing them (question vs just reading the relevant docs).

anna-geller commented:
> > If feasible, can we have these patterns in their relevant documentation sections and just link to them from here? cc @tpdorsey for thoughts.
> 
> I don't mind doing that. The intention was someone asks a question in support and we link them to this one page. I understand the sentiment though that the question might not even be asked in the first place if the info were in the docs. I defer to @tpdorsey to decide. I suspect there will be some that don't fit anywhere. For those, we can leave the answer typed out for now. For any that makes sense somewhere else, we can certainly do that.

I like the idea. When possible, we could have an FAQ section for each feature (e.g. all FAQ items related to run configurations could land in "Run configuration FAQ", similarly to how AWS does it with all their services e.g. [S3 FAQ](https://aws.amazon.com/s3/faqs/)).

kvnkho commented:
As discussed, I will add these stuff to their relevant places in the docs in future PR's. I made an edit to `Is there an equivalent to Airflow sensors in Prefect? How do I trigger event-driven workflows?` and took out `Review` labels from two other questions. I need a re-review now. @tpdorsey

tpdorsey commented:
This looks pretty solid. I think just two additional suggested changes above what I'd suggested previously.

tpdorsey reviewed (dismissed):
Looks good

kvnkho commented:
I think I need a review from you @madkinsz. How do I deal with that failed test?

anna-geller commented:
> I think I need a review from you @madkinsz. How do I deal with that failed test?

@kvnkho rerunning the tests fixed that because the error was just a build VM issue "We had an unexpected error preparing a VM for this build, potentially due to our infrastructure or cloud provider.  Please retry the build in a few minutes"

madkinsz commented:
@kvnkho it looks like there are quite a few unaddressed comments from Terry

kvnkho commented on docs/orchestration/faq/faq.md:
```diff
@@ -0,0 +1,228 @@
+
+### Why do I get `TypeError: cannot pickle ______ object`?
+
+There are two scenarios where this error happens. The first is when using a `DaskExecutor` and using a task input or output that is not serializable by cloudpickle. Dask uses cloudpickle as the mechanism to send data from the client to the workers. This specific error is often raised with mapped tasks that use client type objects, such as connections to databases, as inputs. These connections nee ...
```
I am not sure it's necessarily a best practice because it's the only way things can be done and a Dask limitation. I see the point though. Will reword

kvnkho commented:
> @kvnkho it looks like there are quite a few unaddressed comments from Terry

Ah, my bad I see what happened. I changed one things manually, so the comments were stale and hidden, but they were still there. I went through this again and resolved all of them.

anna-geller commented:
> > @kvnkho it looks like there are quite a few unaddressed comments from Terry
> 
> Ah, my bad I see what happened. I changed one things manually, so the comments were stale and hidden, but they were still there. I went through this again and resolved all of them.

@kvnkho Is there any reason why this hasn't been merged yet? :)

kvnkho commented:
> > > @kvnkho it looks like there are quite a few unaddressed comments from Terry
> > 
> > 
> > Ah, my bad I see what happened. I changed one things manually, so the comments were stale and hidden, but they were still there. I went through this again and resolved all of them.
> 
> @kvnkho Is there any reason why this hasn't been merged yet? :)

I had to update so the review became stale. I request a review again from @madkinsz

madkinsz reviewed (approved):
Relying on Terry's review here.
